File: frappe/utils/legacy_gravatar_cleanup.py
"""One-time cleanup flow for legacy Gravatar URLs stored before Gravatar support was removed.
Can be dropped when skip_gravatar_deletion_prompt is 1 for the majority of sites, or in the
next major release.
"""

from enum import Enum

import frappe
from frappe import _
from frappe.utils import cint
from frappe.utils.background_jobs import is_job_enqueued

GRAVATAR_URL_PATTERN = "%gravatar.com%"
GRAVATAR_DELETION_JOB_ID = "delete_legacy_gravatar_image_urls"
SKIP_GRAVATAR_DELETION_PROMPT = "skip_gravatar_deletion_prompt"
BASE_GRAVATAR_IMAGE_FIELDS = (
	("User", "user_image"),
	("Contact", "image"),
)
ERPNEXT_GRAVATAR_IMAGE_FIELDS = (("Lead", "image"),)


class Action(Enum):
	DELETE_GRAVATAR_URLS = "delete_gravatar_urls"
	KEEP_GRAVATAR_URLS = "keep_gravatar_urls"


def get_gravatar_image_fields():
	fields = list(BASE_GRAVATAR_IMAGE_FIELDS)
	if "erpnext" in frappe.get_installed_apps():
		fields.extend(ERPNEXT_GRAVATAR_IMAGE_FIELDS)
	return fields


def should_show_gravatar_deletion_prompt():
	if "System Manager" not in frappe.get_roles():
		return False

	if cint(frappe.defaults.get_global_default(SKIP_GRAVATAR_DELETION_PROMPT)):
		return False

	if is_job_enqueued(GRAVATAR_DELETION_JOB_ID):
		return False

	return has_gravatar_image_urls()


def has_gravatar_image_urls():
	return any(
		frappe.db.exists(doctype, {fieldname: ("like", GRAVATAR_URL_PATTERN)})
		for doctype, fieldname in get_gravatar_image_fields()
	)


def skip_gravatar_deletion_prompt_if_no_urls():
	if cint(frappe.defaults.get_global_default(SKIP_GRAVATAR_DELETION_PROMPT)):
		return

	if not has_gravatar_image_urls():
		frappe.defaults.set_global_default(SKIP_GRAVATAR_DELETION_PROMPT, 1)


@frappe.whitelist(methods=["POST"])
def submit_gravatar_deletion_prompt(action: str):
	frappe.only_for("System Manager")

	if action == Action.KEEP_GRAVATAR_URLS.value:
		frappe.defaults.set_global_default(SKIP_GRAVATAR_DELETION_PROMPT, 1)
		return "skipped"

	if action == Action.DELETE_GRAVATAR_URLS.value:
		frappe.enqueue(
			delete_gravatar_image_urls,
			queue="long",
			now=frappe.in_test,
			enqueue_after_commit=not frappe.in_test,
			job_id=GRAVATAR_DELETION_JOB_ID,
			deduplicate=True,
		)
		return "queued"

	frappe.throw(_("Invalid action"))


def delete_gravatar_image_urls():
	for doctype, fieldname in get_gravatar_image_fields():
		filters = {fieldname: ("like", GRAVATAR_URL_PATTERN)}
		frappe.db.set_value(doctype, filters, fieldname, "", update_modified=False)

	frappe.defaults.set_global_default(SKIP_GRAVATAR_DELETION_PROMPT, 1)

File: frappe/public/js/frappe/legacy_gravatar_cleanup.js
/* One-time cleanup flow for legacy Gravatar URLs stored before Gravatar support was removed.
 * Can be dropped when skip_gravatar_deletion_prompt is 1 for the majority of sites, or in the
 * next major release.
 */
frappe.provide("frappe.ui");

frappe.ui.maybe_show_legacy_gravatar_cleanup_prompt = function ({ onhide } = {}) {
	if (!frappe.boot.show_gravatar_deletion_prompt) {
		return false;
	}

	const dialog = new frappe.ui.Dialog({
		title: __("Delete Gravatar URLs"),
		fields: [
			{
				fieldname: "message",
				fieldtype: "HTML",
				options: `<p>${__(
					"Frappe no longer uses Gravatar because it can disclose hashed email addresses to a third-party service. To fully stop this behavior, existing Gravatar image URLs must be deleted."
				)} <a href="https://github.com/frappe/frappe/issues/39869" target="_blank">${__(
					"Learn more"
				)}</a></p>
				<p>${
					frappe.utils.get_installed_apps().includes("erpnext")
						? __(
								"Existing Gravatar URLs may still be stored on User, Contact, and Lead records. Do you want to delete these URLs now?"
						  )
						: __(
								"Existing Gravatar URLs may still be stored on User and Contact records. Do you want to delete these URLs now?"
						  )
				}</p>`,
			},
			{
				fieldname: "action",
				label: __("Action"),
				fieldtype: "Select",
				options: [
					"",
					{ label: __("Delete Gravatar URLs"), value: "delete_gravatar_urls" },
					{ label: __("Keep Gravatar URLs"), value: "keep_gravatar_urls" },
				],
				reqd: 1,
			},
		],
		primary_action_label: __("Confirm", null, "Confirm gravatar deletion prompt"),
		primary_action: ({ action }) => {
			return frappe
				.xcall("frappe.utils.legacy_gravatar_cleanup.submit_gravatar_deletion_prompt", {
					action,
				})
				.then((message) => {
					if (message === "queued") {
						frappe.show_alert({
							message: __("Gravatar URL deletion has been queued."),
							indicator: "blue",
						});
					} else if (message === "skipped") {
						frappe.show_alert({
							message: __("Gravatar URLs will not be deleted."),
							indicator: "green",
						});
					}

					frappe.boot.show_gravatar_deletion_prompt = false;
				})
				.finally(() => dialog.hide());
		},
	});

	dialog.onhide = onhide;
	dialog.show();
	return true;
};
